docs/cheat-sheets/python-command-injection.mdx

```diff
@@ -0,0 +1,649 @@
+
+Example:
+
+```python
+import asyncio
+
+# Vulnerable
+user_input = "cat /etc/passwd" # value supplied by user
+loop = asyncio.new_event_loop()
+# This is similar to the standard library subprocess.Popen class called with shell=True
+loop.subprocess_shell(asyncio.SubprocessProtocol, user_input)
+
+# Vulnerable
+user_input = "cat /etc/passwd" # value supplied by user
+asyncio.subprocess.create_subprocess_shell(user_input)
+```
+
+#### References:
+
+- [**asyncio subprocess** documentation](https://docs.python.org/3/library/asyncio-subprocess.html)
+- [**Event Loop subprocess_shell** documentation](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.subprocess_shell)
+
+#### Mitigation
+
+Do not let user input in `asyncio subprocess` methods. Alternatively,
+
+- Always try to use internal Python API (if it exists) instead of running an OS command.
+- Consider using asyncio.subprocess functions with array of program arguments (e.g. `create_subprocess_exec`).
+- Don’t pass user-controlled input.
+- If it is not possible, then do not let running arbitrary commands, use a white list for inputs.
+
+#### Semgrep rule
+
+<LinkToRegistryRule ruleId="python.lang.security.audit.dangerous-asyncio-shell.dangerous-asyncio-shell" />
+
+### 1.H. Creating subprocesses asynchronously
+
+`asyncio.subprocess` also allows to create subprocesses asynchronously.
+Such methods as `create_subprocess_exec` and Event Loop's `subprocess_exec` are intended for creating a subprocess from one or more string arguments specified by args.
+Allowing user input in a command that is passed as an argument to one of these methods, can create an opportunity for a Command Injection vulnerability.
+
+Example:
+
+```python
+import asyncio
+
+# Vulnerable
+user_input = "/evil/exe" # value supplied by user
+loop = asyncio.new_event_loop()
+loop.subprocess_exec(asyncio.SubprocessProtocol, [user_input, "--parameter"])
+
+# Vulnerable
+user_input = "cat /etc/passwd" # value supplied by user
+asyncio.subprocess.create_subprocess_exec("bash", ["bash", "-c", user_input])
+
+# Not vulnerable
+user_input = "/evil/exe" # value supplied by user
+loop = asyncio.new_event_loop()
+loop.subprocess_exec(asyncio.SubprocessProtocol, ['ls', '-l', user_input])
+```
+
+#### References:
+
+- [**asyncio subprocess** documentation](https://docs.python.org/3/library/asyncio-subprocess.html)
+
+#### Mitigation
+
+Do not let user input in `asyncio subprocess` methods. Alternatively,
+
+- Always try to use internal Python API (if it exists) instead of running an OS command.
+- Don’t pass user-controlled input.
+- Use `shlex.split` to correctly parse command string into an array and `shlex.quote` to correctly sanitize input as a command-line parameter.
+- Avoid running `sh` as a command with arguments, if it is not possible - strip everything except alphanumeric characters from an input provided for the command string and arguments.
```

Comment:
```suggestion
- Avoid running `sh` as a command with arguments. If it’s not possible to avoid, strip everything except alphanumeric characters from an input provided for the command string and arguments.
```